format/nes/nes.go

```diff
@@ -0,0 +1,325 @@
+
+	if nes20 == 0b10 { // NES 2.0
+		// byte 7 second half and byte 8
+		mapperMSB := d.FieldU8("mapper_msb")
+		d.FieldValueUint("mapper_full", mapperLSB+(mapperMSB<<4))
+		d.FieldU4("submapper")
+
+		// byte 9
+		prgMSB := d.FieldU4("prg_rom_size_msb")
+		chrMSB := d.FieldU4("chr_rom_size_msb")
+		prgROMSize := romSize(prgLSB, prgMSB, 16)
+		d.FieldValueUint("prg_rom_size_full", prgROMSize)
+		chrROMSize := romSize(chrLSB, chrMSB, 8)
+		d.FieldValueUint("chr_rom_size_full", chrROMSize)
+
+		// byte 10
+		d.FieldU4("prg_ram_size", shiftSizeMapper)
+		d.FieldU4("prg_nvram_size", shiftSizeMapper)
+
+		// byte 11
+		d.FieldU4("chr_ram_size", shiftSizeMapper)
+		d.FieldU4("chr_nvram_size", shiftSizeMapper)
+
+		// byte 12
+		d.FieldU2("cpu_ppu_timing_mode", timingModeMapper)
+		d.SeekRel(6)
+
+		// byte 13
+		switch consoleType {
+		case 1:
+			d.FieldU4("vs_ppu_type", vsPPUMapper)
+			d.FieldU4("vs_hardware_type", vsHardwareMapper)
+		case 3:
+			d.FieldU4("ext_console_type", extConsoleMapper)
+			d.SeekRel(4)
+		default:
+			d.SeekRel(8)
+		}
+
+		// byte 14
+		d.FieldU2("misc_roms")
+		d.SeekRel(6)
+
+		// byte 15
+		d.FieldU6("default_exp_device")
+		d.SeekRel(2)
+
+		nc.prgROMSize = prgROMSize
+		nc.chrROMSize = chrROMSize
+	} else { // iNES
+		// byte 7 (second half)
+		mapperMSB := d.FieldU4("mapper_msb")
+		d.FieldValueUint("mapper_full", mapperLSB+(mapperMSB<<4))
+
+		prgROMSize := romSize(prgLSB, 0, 16)
+		d.FieldValueUint("prg_rom_size_full", prgROMSize)
+		chrROMSize := romSize(chrLSB, 0, 8)
+		d.FieldValueUint("chr_rom_size_full", chrROMSize)
+
+		// byte 8
+		d.FieldU8("prg_ram_size", ramSizer(8))
+
+		// byte 9 - 15
+		d.FieldStr("crap", 7, encoding.Nop)
+
+		nc.prgROMSize = prgROMSize
+		nc.chrROMSize = chrROMSize
+	}
+
+	nc.nes20 = nes20 == 0b10
+
+	if trainer {
+		nc.trainerSize = 512
+	} else {
+		nc.trainerSize = 0
+	}
+
+}
+
+func decodePRGROM(d *decode.D) {
+	for !d.End() {
+		peek := d.PeekUintBits(8)
+		peekTyp := opMap[peek].Type
+		peekArgL := ArgLength(peekTyp)
+
+		if d.BitsLeft() < int64(1+peekArgL)*8 {
+			d.FieldRawLen("junk", d.BitsLeft())
+		} else {
+			d.FieldStruct("instruction", func(d *decode.D) {
+				op := d.FieldU8("op_code", scalar.UintFn(func(s scalar.Uint) (scalar.Uint, error) {
+					opLookup := opMap[s.Actual]
+					s.Sym = opLookup.Name
+
+					return s, nil
+				}))
+
+				typ := opMap[op].Type
+				argL := ArgLength(typ)
+				switch argL {
+				case 1:
+					d.FieldU8("args", GetArgFormatter(typ))
+				case 2:
+					d.FieldU16("args", GetArgFormatter(typ))
+				}
+			})
+		}
+	}
+}
+
+func reverse(s string) string {
+	runes := []rune(s)
+	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
+		runes[i], runes[j] = runes[j], runes[i]
+	}
+	return string(runes)
+}
+
+func decodeTilePart(d *decode.D) scalar.Uint {
+	a, err := d.TryU64()
+	if err != nil {
+		panic(decode.IOError{Err: err, Name: "tile", Op: "Uint", Pos: d.Pos()})
+	}
```

Comment:
You want to throw tile specific panic error? if you do `a := d.U64()` it will also panic behind the scene. If you want custom fatal error you can use `d.Fatalf` or for non-fatal (can be skipped with `-o force=true`) use `d.Errorf`,

Reply:
Yeah, this looks like late night brain fog to me. =) I must have been deep into decode_gen.go when I wrote this.

Reply:
No worries :) i've lost count of times i've done something but then realised that i forgot i had already prepared or fix something month ago